wgsl/index.bs

```diff
@@ -2544,51 +2547,103 @@ programmable parts of pipelines:
 
-The `entry_point` declares an entry point into the module. The entry points may
-be forward declarations but the functions referenced must be declared in the
-file.
+An <dfn noexport>entry point</dfn> is a function that is invoked to perform
+the work for a particular [=shader stage=].
 
-The input and output parameters to the entry point are determined by which
-global variables are used in the function and any called functions.
+Specify a `stage` attribute on a function declaration to declare that function
+as an entry point.
 
-The entry point function must be declared with no arguments and returning `void`.
+When configuring the stage in the pipeline, the entry point is specified by providing
+the [SHORTNAME] module and the entry point's function name.
 
-<pre class='def'>
-entry_point_decl
-  : ENTRY_POINT pipeline_stage EQUAL IDENT
-  | ENTRY_POINT pipeline_stage AS STRING_LITERAL EQUAL IDENT
-  | ENTRY_POINT pipeline_stage AS IDENT EQUAL IDENT
-</pre>
+An entry point function must have no parameters,
+and its return type must be [=void=].
+
+ISSUE: We've lost the ability to use shader names that are invalid identifiers.
+Do Metal and D3D12 limit entry point names to valid identifiers in C?
 
 <div class='example' heading='Entry Point'>
   <xmp>
-    entry_point vertex = main
-       OpEntryPoint Vertex %vtx_main "vtx_main" %gl_FragColor
+    [[stage vertex]]
+    fn vtx_main() -> void { gl_Position = vec4<f32>(); }
+       # OpEntryPoint Vertex %vtx_main "vtx_main" %gl_Position
 
-    entry_point fragment as “frag_main” = main
-       OpEntryPoint Fragment %main "frag_main" %gl_FragColor
+    [[stage fragment]]
+    fn frag_main -> void { gl_FragColor = vec4<f32>(); }
+       # OpEntryPoint Fragment %frag_main "frag_main" %gl_FragColor
 
-    entry_point compute = comp_main
-       OpEntryPoint GLCompute %comp_main "comp_main" %gl_FragColor
+    [[stage compute]]
+    fn main() -> void { }
+       # OpEntryPoint GLCompute %main "main"
   </xmp>
 </div>
+
+The set of <dfn noexport>functions in a shader stage</dfn> is the union of:
+
+* The entry point function for the stage.
+* The targets of function calls from within the body of a function
+    in the shader stage, whether or not that call is executed.
+
+The union is applied repeatedly until it stabilizes.
+It will stabilize in a finite number of steps.
+
+### Function attributes for entry points ### {#entry-point-attributes}
+
+<pre class='def'>
+function_decoration
+  : STAGE pipeline_stage
+  | WORKGROUP_SIZE INT_LITERAL (INT_LITERAL INT_LITERAL?)?
+
+pipeline_stage
+  : COMPUTE
+  | VERTEX
+  | FRAGMENT
+</pre>
+
+: <dfn noexport>stage</dfn>
+:: The `stage` attribute declares that a function is an entry point for particular pipeline stage.
+: <dfn noexport>workgroup_size</dfn>
+:: The `workgroup_size` attribute specifies the x, y, and z dimensions of the [=workgroup grid=]
+    for a [=compute=] shader.
+    The size in the x dimension is provided by the first literal.
```

Comment:
This will default to 1 as well if the workgroup_size is left off. Should we note that here?

Reply:
I didn't put that here because you can't write a workgroup_size attribute without the first dimension.

However, in the later section we do say that if no workgroup_size attribute is specified, then (1,1,1) is assumed.